voxel_grid/prototype/mvp_v2.hpp

```diff
@@ -82,9 +82,15 @@ class TransformFilter : public Filter<PointT> {
         output.push_back(res.value());
       }
     }
+  }
+
+  // need to declare the corresponding functions in GridStruct
+  std::vector<int> getLeafLayout() const {
+    return grid_struct_.getLeafLayout();
+  }
 
-    // push back leaf_layout_ to the base
-    grid_struct_.saveResults(this);
+  inline int getCentroidIndex(float x, float y, float z) const {
+    return grid_struct_.getCentroidIndex(x, y, z);
```

Comment:
Can we move these to GridFilterBase and instead replace them with `getTransformer` or similar to return reference to `grid_struct_`?

Reply:
Yes, we can have a function returning the reference.

But we have a few functions which are specific to certain filters, like `getCentroidIndex` is only for `VoxelGrid` and `VoxelGridLabel`. (we have 5 grid filters in total) So adding those functions to either `GridFilterBase` or `TransformFilter` I think is implausible.
If we want to keep APIs of `VoxelGrid` the same as before, the only way I can think of is to add another class to declare those filter specific functions. 

@kunaltyagi Do you have any better idea to get around that?

Reply:
Those functions can go in the last tier (transform->gridBase->VoxelFilter) right?

Reply:
Done. I created a new class as `VoxelFilter` and `VoxelGrid` is once again declared by `using`

## Brief recap for myself

### Goal of this prototype

> Pro
>     - Prevents multiple inheritance
>     - Restricts GridStruct from adding any API/modifying functionality
>
>  Cons
>     - Requires used to sub-class to add functionality

Seems like we achieved both of the pro

### Initial goal for grid filters

  [-] Reduce boilerplate code
  [✓] Allow users pass their `GridStruct` to have custom grid filter
  [✓] Allow upgrading concurrence with ease

### Comments
1. Reduce boilerplate code
    - [✓]  boilerplate logic in `applyFilter` is removed
    - [✕] Not all common code is reduced
        - Setting up `min_b_` and `max_b_` are common, but they are just a few lines of code (usually >10 lines)
        - Hashing is basically universal -> **TODO**: move to `GridFilterBase`
 2. Allow users pass their `GridStruct` to have custom grid filter
    - [✓] Users can pass their `GridStruct` to `TransformFilter`
 3. Allow upgrading concurrence with ease
     - [✓] All for loops are located in `applyFilter` now, all grid filters including filters custom `GridStruct` can be benefit from concurrence